File: pathfinder-bukkit/src/main/java/de/cubbossa/pathfinder/command/FindCommand.java
package de.cubbossa.pathfinder.command;

import de.cubbossa.pathfinder.PathFinder;
import de.cubbossa.pathfinder.PathPerms;
import de.cubbossa.pathfinder.graph.GraphEntryNotEstablishedException;
import de.cubbossa.pathfinder.graph.NoPathFoundException;
import de.cubbossa.pathfinder.messages.Messages;
import de.cubbossa.pathfinder.misc.PathPlayer;
import de.cubbossa.pathfinder.navigation.NavigationLocation;
import de.cubbossa.pathfinder.navigation.NavigationModule;
import de.cubbossa.pathfinder.navigation.Route;
import de.cubbossa.pathfinder.node.NodeSelectionImpl;
import de.cubbossa.pathfinder.node.implementation.PlayerNode;
import dev.jorel.commandapi.CommandTree;
import java.util.concurrent.CompletionException;
import java.util.logging.Level;
import org.bukkit.entity.Player;

public class FindCommand extends CommandTree {

  public FindCommand() {
    super("find");
    withAliases("gps", "navigate");
    withPermission(PathPerms.PERM_CMD_FIND);

    then(Arguments.navigateSelectionArgument("selection")
        .executesPlayer((player, args) -> {
          NodeSelectionImpl targets = args.getUnchecked(0);
          if (targets == null) {
            return;
          }
          PathPlayer<Player> p = PathPlayer.wrap(player);
          if (targets.isEmpty()) {
            p.sendMessage(Messages.CMD_FIND_EMPTY);
            return;
          }
          NavigationModule<Player> navigationModule = NavigationModule.get();
          navigationModule.setFindCommandPath(p, Route
                  .from(NavigationLocation.movingExternalNode(new PlayerNode(p)))
                  .toAny(targets))
              .whenComplete((nav, throwable) -> {
                if (throwable != null) {
                  if (throwable instanceof CompletionException) {
                    throwable = throwable.getCause();
                  }
                  if (throwable instanceof NoPathFoundException) {
                    p.sendMessage(Messages.CMD_FIND_BLOCKED);
                  } else if (throwable instanceof GraphEntryNotEstablishedException) {
                    p.sendMessage(Messages.CMD_FIND_TOO_FAR);
                  } else {
                    p.sendMessage(Messages.CMD_FIND_UNKNOWN);
                    PathFinder.get().getLogger().log(Level.SEVERE, "Unknown error while finding path.", throwable);
                  }
                  return;
                }
                nav.persist().cancelWhenTargetInRange();
              });
        })
    );
  }
}

File: pathfinder-bukkit/src/main/java/de/cubbossa/pathfinder/command/FindCommand.kt
package de.cubbossa.pathfinder.command

import de.cubbossa.pathfinder.PathFinder
import de.cubbossa.pathfinder.PathPerms
import de.cubbossa.pathfinder.asPathPlayer
import de.cubbossa.pathfinder.graph.GraphEntryNotEstablishedException
import de.cubbossa.pathfinder.graph.NoPathFoundException
import de.cubbossa.pathfinder.launchIO
import de.cubbossa.pathfinder.messages.Messages
import de.cubbossa.pathfinder.misc.PathPlayer
import de.cubbossa.pathfinder.navigation.NavigationLocation
import de.cubbossa.pathfinder.navigation.NavigationModule.Companion.get
import de.cubbossa.pathfinder.navigation.Route
import de.cubbossa.pathfinder.node.NodeSelectionImpl
import de.cubbossa.pathfinder.node.implementation.PlayerNode
import dev.jorel.commandapi.CommandTree
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import org.bukkit.entity.Player
import java.util.logging.Level

class FindCommand : CommandTree("find") {
    init {
        withAliases("gps", "navigate")
        withPermission(PathPerms.PERM_CMD_FIND)

        navigateSelectionArgument("selection") {
            executesPlayer(PlayerCommandExecutor { player, args ->
                val targets: NodeSelectionImpl =
                    args.getUnchecked(0) ?: return@PlayerCommandExecutor
                val p: PathPlayer<Player> = player.asPathPlayer()
                if (targets.isEmpty()) {
                    p.sendMessage(Messages.CMD_FIND_EMPTY)
                    return@PlayerCommandExecutor
                }
                launchIO {
                    val navigationModule = get<Player>()
                    try {
                        val path = navigationModule.navigate(
                            p, Route
                                .from(NavigationLocation.movingExternalNode(PlayerNode(p)))
                                .toAny(targets)
                        )
                        navigationModule.cancelPathWhenTargetReached(path)
                    } catch (t: NoPathFoundException) {
                        p.sendMessage(Messages.CMD_FIND_BLOCKED)
                    } catch (t: GraphEntryNotEstablishedException) {
                        p.sendMessage(Messages.CMD_FIND_TOO_FAR)
                    } catch (t: Throwable) {
                        p.sendMessage(Messages.CMD_FIND_UNKNOWN)
                        PathFinder.get().logger.log(
                            Level.SEVERE,
                            "Unknown error while finding path.",
                            t
                        )
                    }
                }
            })
        }
    }
}
